README.md
# andreasklein_org

**.. My personal space for article, notes & ideas on the web. Written in pure Javascript powered by [Node.js](http://nodejs.org/), [Grunt.js](http://gruntjs.com/) & [Express](http://expressjs.com/)**


## Setup Development Enviroment
### System Requirements

* node.js (>= v0.10.x)
* git (>= 1.7.x)

### Registered Grunt Tasks

__1) `default`__<br>
Default Task used for dev workflow. Runs `$ grunt dev` Task.

__2) `dev`__<br>
Development Workflow

* `jshint` Javascript Code-linting
* `less` Compiles LESS files
* `autoprefixer` Completes CSS vendor prefixes automatically.
* `concat` Concatenating Javascript files into one single file.
* `connect` Starts Node.js Webserver.
* `open` Opens the dev environment in Google Chrome
* `watch` Watches your working dir and live-reloads the browser window after making changes

__3) `prod`__<br>
Runs a distribution build.
* Includes all `dev` tasks, expect `connect` & `watch` tasks
* `uglify` Javascript minifcation + banner
* `cssmin` CSS minification + banner

__4) `bump`__<br>
Bump package version, create tag, commit, push...
Uses [grunt-bump](https://github.com/vojtajina/grunt-bump) task.

* `$ grunt bump:patch` // 0.0.1 to 0.0.2
* `$ grunt bump:minor` // 0.1.0 to 0.2.0
* `$ grunt bump:major` // 0.x to 1.0.0
* `$ grunt bump:build` // 1.0.0 to 1.0.0-1

&copy; 2015 Andreas Klein


CHANGELOG.md
# Changelog

## HEAD
* Use Gravatar for Profile Image
* Update global text line-height
* Add some semantic flavour to article pages :)

## v1.3.1 / 2015-03-05
* Remove CSS IDs and replace with classes
* Some minor CSS Refactoring
* Move to REMs instead of EMs
* Update Copyright date to 2015
* Update meta title
* Add banner option for CSS files to grunt-contrib-less

## v1.3.0 / 2014-10-20
* Move contents directory to separate git repo and fetch it as git-submodule: https://github.com/ixisio/andreasklein_org-contents
* Update npm dependencies to latest version
* Add JavaScript Code Style Checker (JSCS)
* Refactor Code Style
* Clean up some less styles indentation
* Add social profile link rel to <head>
* Add indieweb Authorization endpoints
* Remove express-device
* Rename main to core.css
* Remove mobile avatar image
* Remove Open-Sanse true-type font format
* Remove icons eot format and remove .eot file
* Remove underscore.js from npm depedencies
* Migrate jade to v1.7.x
* Migrate express to v4.x
* Remove mobile stylesheet and text-image-block elements

## v1.2.0 / 2014-06-19
* Performance optimization
* Update Express Caching
* Desktop CSS out sourcing
* Update about-me section

## v1.1.2 / 2014-05-30
* Performance optimization

## v1.1.1 / 2014-05-30
* Patch: Tablets up to 1024 Pixels has no styling! (fixed)

## v1.1.0 / 2014-05-29
* Update dependencies to latest version
* Add grunt tasks documentation
* Move from SASS to LESS.js for CSS precompiling
* Media-Query outsourcing (experiment with RESS)
* Some minor changes & improvements

## v1.0.0 / 2014-02-11
* update / migrate poet to v1.0.0-rc4
* Fix some a11y issues
* add rss link
* some minor bugfixes & optimizations

## v0.12.0 / 2013-12-06
* Add a new fresh set of Icons
* Clean up SASS Files
* Change the look of `<h2>` Tags
* Visual changes to 404 page
* Update grunt.js to v0.4.2
* Some minor improvements

## v0.11.0 / 2013-10-24
* Frameless figures on phones. Image indentation on larger screen.
* Styling figure captions.
* Refactoring

## v0.10.1 / 2013-10-16
* Enable :hover states only for non-touch devices


## v0.10.0 / 2013-10-16
* Update prettyPrint.js to parse all `@code <pre>` tags, instead of looking for className `.prettyprint`.
* Update 404 Templates
  * Send right 404 Status. With Express Middleware this looks like `res.status(404);`
  * Update 404 Page Texts
* Performace optimization
  * Remove `v7.js` & the whole jQuery library
  * Start prettyPrint() by body.onload event
  * No Email-Address obfuscating anymore

## v0.9.2 / 2013-09-26
* Remove :hover Styles for touch-enabled devices

## v0.9.1 / 2013-09-26
* Remove headings uppercase styling
* Add matchdep for grunt
* add grunt-contrib-open

## v0.9.0 / 2013-05-31
* Make some _type CSS changes
* Article Page Styling
* Remove Maintenance Handler/Mode

## v0.8.0
* Add 404 Page Route & Handler
* New Template/Layout Structure
* Add Category & Tag Routes
* Styling posts, tags

## v0.6.0
* Add RSS Route
* Add RSS Handler

## v0.4.0
* Finish mobile breakpoint styling
* Add Maintenance Screen
* Add Google Analytics

## v0.2.0
* Add Heroku Procfile
* Add Imprint Page Route & Content

## v0.1.0
* Add grunt build script (server, compass, lint, watch)
* finish layout styling overview pages
* add poet
* add express
* add basic routes (pages, posts, cats, tags)
* finish homepage


blog.js
var zlib      = require('zlib'),
    express   = require('express'),
    device    = require('express-device'),
    app       = express(),
    Poet      = require('poet'),
    readCSS   = require('./lib/read-css');

function getCssPerView(view, cb) {
  if (view === 'desktop' || view === 'tablet') {
    readCSS('www/styles/views/desktop.css', function(css) {
      cb(css);
    });
  } else {
    cb(false);
  }
}

// Enable express-device helper
// https://github.com/rguerreiro/express-device
app.use(device.capture());
app.use(express.compress());
app.use(function(req, res, next) {
  getCssPerView(req.device.type, function(css) {
    req.css = css;
    next();
  });
});

// Instantiate and hook Poet into express;
// ---------------------
 var poet = Poet(app, {
  postsPerPage: 10,
  posts: './content/',
  metaFormat: 'json',
  readMoreLink : function ( post ) {
     var anchor = '<a href="'+post.url+'" title="Read more of '+post.title+'">&rsaquo; read more</a>';
     return '<p class="readmore">' + anchor + '</p>';
  },
  routes: {
    '/articles/:post': 'post',
    '/tags/:tag': 'tag'
  }
});


// Express Config
// ---------------------

// Prettify HTML
//app.locals.pretty = true;
app.set( 'view engine', 'jade' );
app.set('view options', { layout: false }); // express-device
app.set( 'views', __dirname + '/views' );
app.use( express.static( __dirname + '/www' ));
app.use( app.router );


// Poet Init
// ---------------------
poet.init().then(function () {
  poet.clearCache();
});


// Poet Routes
// ---------------------
poet.addRoute( '/articles/:post', function ( req, res ) {
  var post = poet.helpers.getPost( req.params.post );

  if ( post ) {
    res.render( 'post', {
      post: post,
      device: req.device,
      css: req.css
    });
  } else {
    res.status(404);
    res.render('4o4.jade', {
      url: req.url,
      device: req.device,
      css: req.css
    });
  }
});

poet.addRoute( '/tag/:tag', function ( req, res ) {
  var taggedPosts = poet.helpers.postsWithTag( req.params.tag );
  if ( taggedPosts.length ) {
    res.render( 'tag', {
      posts : taggedPosts,
      tag : req.params.tag,
      device: req.device,
      css: req.css
    });
  }
});


// Static Routes
// ----------------
app.get('/legal', function ( req, res ) {
  res.render( 'page-legal', {
    title: 'Imprint',
    device: req.device,
    css: req.css
  });
});

app.get('/rss', function (req, res) {
  var posts = poet.helpers.getPosts(0, 5);
  res.setHeader('Content-Type', 'application/rss+xml');
  res.render('rss', {
    posts: posts,
    device: req.device,
    css: req.css
  });
});

app.get( '/', function ( req, res ) {
  res.render( 'index', {
    device: req.device,
    css: req.css
  });
});

app.use(function(req, res) {
  res.status(404);
  res.render('4o4.jade', {
    url: req.url,
    device: req.device,
    css: req.css
  });
});

app.listen( process.env.PORT || 3322 );
console.log('Listen on Port ' + (process.env.PORT || 3322) );


index.js
var express = require('express'),
    app = express(),
    Poet = require('poet');

/**
 * Instantiate and hook Poet into Express.js
 */
 var poet = Poet(app, {
    postsPerPage: 10,
    posts: './content/posts/',
    notes: './content/notes/',
    metaFormat: 'yaml',
    readMoreLink: function (post) {
       var anchor = '<a href="' + post.url + '" title="Read more of ' + post.title + '">&rsaquo; read more</a>';

       return '<p class="readmore">' + anchor + '</p>';
    },
    routes: {
        '/articles/:post': 'post',
        '/notes/:note': 'note',
        '/tags/:tag': 'tag'
    }
});

/**
 * Configure Express.js
 */
app.set('view engine', 'jade');
app.set('view options', { layout: false });
app.set('views', __dirname + '/views');
app.use(express.static(__dirname + '/www'));

/**
 * Initialize poet and clear poet cache
 *
 * @return {void}
 */
poet.init().then(function () {
  poet.clearCache();
});

/**
 * Route Posts | Render article page if exists, otherwise 404
 *
 * @param  {Object} req Poet request parameter
 * @param  {Object} res Poet response parameter
 * @return {void}
 */
poet.addRoute('/articles/:post', function (req, res) {
  var post = poet.helpers.getPost(req.params.post);

  if (post) {
      res.render('post', {
          post: post,
          css: req.css
      });
  } else {
      res.status(404);
      res.render('4o4.jade', {
          url: req.url,
          css: req.css
      });
  }
});

/**
 * Route Note | Render note page if exists, otherwise 404
 *
 * @param  {Object} req Poet request parameter
 * @param  {Object} res Poet response parameter
 * @return {void}
 */
poet.addRoute('/notes/:note', function (req, res) {
  var note = poet.helpers.getPost(req.params.post);

  if (post) {
      res.render('post', {
          post: post
      });
  } else {
      res.status(404);
      res.render('4o4.jade', {
          url: req.url
      });
  }
});

/**
 * Route Tags | Render page with tagged posts
 *
 * @param  {Object} req Poet request parameter
 * @param  {Object} res Poet response parameter
 * @return {void}
 */
poet.addRoute('/tag/:tag', function (req, res) {
    var taggedPosts = poet.helpers.postsWithTag(req.params.tag);

    if (taggedPosts.length) {
        res.render('tag', {
            posts: taggedPosts,
            tag: req.params.tag
        });
    }
});

/**
 * Route Legal | Render imprint page
 *
 * @param  {Object} req   Express request parameter
 * @param  {Object} res   Express response parameter
 * @return {void}
 */
app.get('/legal', function (req, res) {
    res.render('page-legal', {
        title: 'Imprint'
    });
});

/**
 * Route RSS | Render 5 latest posts
 *
 * @param  {Object} req   Express request parameter
 * @param  {Object} res   Express response parameter
 * @return {void}
 */
app.get('/rss', function (req, res) {
    var posts = poet.helpers.getPosts(0, 5);

    res.setHeader('Content-Type', 'application/rss+xml');
    res.render('rss', {
        posts: posts
    });
});

/**
 * Route Startpage
 *
 * @param  {Object} req   Express request parameter
 * @param  {Object} res   Express response parameter
 * @return {void}
 */
app.get('/', function (req, res) {
    res.render('index');
});

/**
 * Route About
 *
 * CURRENTLY NOT YET ENABLED !!
 *
 * @param  {Object} req   Express request parameter
 * @param  {Object} res   Express response parameter
 * @return {void}
 */
// app.get( '/about', function (req, res) {
//   res.render( 'page-about', {
//     title: 'About me',
//     css: req.css
//   });
// });
//

/**
 * Route 404
 *
 * @param  {Object} req   Express request parameter
 * @param  {Object} res   Express response parameter
 * @return {void}
 */
app.use(function (req, res) {
    res.status(404);
    res.render('4o4.jade', {
        url: req.url
    });
});

app.listen(process.env.PORT || 3322);


Gruntfile.js
/**
 * Gruntfile
 *
 * @author Andreas Klein
 * @date 2013-05-26
 */
module.exports = function (grunt) {
  'use strict';

  require('matchdep').filterDev('grunt-*').forEach(grunt.loadNpmTasks);

  grunt.initConfig({
    globals: {
        DEV_SERVER_PORT: 3322
    },
    banner: '/*! <%= pkg.name %> - v<%= pkg.version %>- ' +
            '<%= grunt.template.today("yyyy-mm-dd") %> */',
    pkg: grunt.file.readJSON('package.json'),
    jshint: {
        options: {
            curly: true,
            eqeqeq: true,
            eqnull: true,
            browser: true,
            globals: {
                jQuery: true
            }
        },
        src: ['*.js', 'www/scripts/*.js']
    },
    jscs: {
        src: '<%= jshint.src %>',
        options: {
            config: '.jscsrc'
        }
    },
    less: {
        main: {
            options: {
                compress: false,
                report: false,
                dumpLineNumbers: 'comments',
                banner: '<%= banner %>'
            },
            files: {
                'www/styles/core.css': 'www/styles/less/core.less'
            }
        }
    },
    autoprefixer: {
        options: {
            browsers: ['last 1 version'],
            report: 'min'
        },
        css: {
            src: ['www/styles/**/*.css']
        }
    },
    watch: {
        less: {
            files: ['www/styles/**/*.less'],
            tasks: ['less', 'autoprefixer']
        },
        js: {
            files: ['*.js', 'www/scripts/*.js'],
            tasks: ['jshint', 'jscs']
        }
    },
    open: {
        server: {
            path: 'http://localhost:<%= globals.DEV_SERVER_PORT %>'
        }
    },
    cssmin: {
        options: {
            report: 'min'
        },
        dist: {
            expand: true,
            cwd: 'www/styles/',
            src: ['**/*.css'],
            dest: 'www/styles/'
        }
    }
  });

  /**
   * Register Tasks:
   *
   *   (1) 'prod': Build for Prod enviroment
   *   (2) 'assets' : Lint & build static assets like (SASS Files)
   *   (3) 'dev' Development Workflow
   *   (4) default (run `dev` task)
   */
  grunt.registerTask('prod', [
      'jshint',
      'jscs',
      'less',
      'autoprefixer',
      'cssmin'
  ]);

  grunt.registerTask('assets', [
      'jshint',
      'jscs',
      'autoprefixer'
  ]);

  grunt.registerTask('node-server', function () {
      grunt.util.spawn({
          cmd: 'node',
          args: ['index.js']
      });
  });

  grunt.registerTask('dev', function () {
      grunt.task.run('node-server');
      grunt.task.run('assets');
      grunt.task.run('open');
      grunt.task.run('watch');
  });

  grunt.registerTask('default', ['dev']);
};
